internal/pkg/deploy/cloudformation/stack/transformers.go

```diff
@@ -124,3 +150,149 @@ func logConfigOpts(lc *manifest.Logging) *template.LogConfigOpts {
+	}, nil
+}
+
+// renderSidecarMountPoints is used to convert from manifest to template objects.
+func renderSidecarMountPoints(in []manifest.SidecarMountPoint) ([]*template.MountPoint, error) {
+	if len(in) == 0 {
+		return nil, nil
+	}
+	output := []*template.MountPoint{}
+	for _, smp := range in {
+		if smp.ContainerPath == nil {
+			return nil, errMPNoContainerPath
+		}
+		if aws.StringValue(smp.SourceVolume) == "" {
+			return nil, errNoSourceVolume
+		}
+		readOnly := aws.Bool(true)
+		if smp.ReadOnly != nil {
+			readOnly = smp.ReadOnly
+		}
+		mp := template.MountPoint{
+			ContainerPath: smp.ContainerPath,
+			SourceVolume:  smp.SourceVolume,
+			ReadOnly:      readOnly,
+		}
+		output = append(output, &mp)
+	}
+	return output, nil
+}
+
+func renderStoragePermissions(input map[string]manifest.Volume) ([]*template.EFSPermission, error) {
+	if len(input) == 0 {
+		return nil, nil
+	}
+	output := []*template.EFSPermission{}
+	for _, volume := range input {
+		// Write defaults to false
+		write := defaultWritePermission
+		if volume.ReadOnly != nil {
+			write = !aws.BoolValue(volume.ReadOnly)
+		}
+		if volume.EFS.FileSystemID == nil {
+			return nil, errNoFSID
+		}
+		perm := template.EFSPermission{
+			Write:         write,
+			AccessPointID: volume.EFS.AuthConfig.AccessPointID,
+			FilesystemID:  volume.EFS.FileSystemID,
+		}
+		output = append(output, &perm)
+	}
+	return output, nil
+}
+
+func renderMountPoints(input map[string]manifest.Volume) ([]*template.MountPoint, error) {
+	if len(input) == 0 {
+		return nil, nil
+	}
+	output := []*template.MountPoint{}
+	for name, volume := range input {
+		// ContainerPath must be specified.
+		if volume.ContainerPath == nil {
+			return nil, errVolNoContainerPath
+		}
+		// ReadOnly defaults to true.
+		readOnly := defaultReadOnly
+		if volume.ReadOnly != nil {
+			readOnly = volume.ReadOnly
+		}
+		mp := template.MountPoint{
+			ReadOnly:      readOnly,
+			ContainerPath: volume.ContainerPath,
+			SourceVolume:  aws.String(name),
+		}
+		output = append(output, &mp)
+	}
+	return output, nil
+}
+
+func renderVolumes(input map[string]manifest.Volume) ([]*template.Volume, error) {
+	if len(input) == 0 {
+		return nil, nil
+	}
+	output := []*template.Volume{}
+	for name, volume := range input {
+		// Set default values correctly.
+		fsID := volume.EFS.FileSystemID
+		if aws.StringValue(fsID) == "" {
+			return nil, errNoFSID
+		}
+		rootDir := volume.EFS.RootDirectory
+		if aws.StringValue(rootDir) == "" {
+			rootDir = defaultRootDirectory
+		}
+		var iam *string
+		if volume.EFS.AuthConfig.IAM == nil {
+			iam = defaultIAM
+		}
+		if aws.BoolValue(volume.EFS.AuthConfig.IAM) {
+			iam = pEnabled
+		}
+
+		// Validate ECS requirements: when an AP is specified, IAM MUST be true
+		// and root directory MUST be either empty or "/"
+		if aws.StringValue(volume.EFS.AuthConfig.AccessPointID) != "" {
+			if !aws.BoolValue(volume.EFS.AuthConfig.IAM) {
+				return nil, errAccessPointWithoutIAM
```

Comment:
I like the fact that for complicated logic we can do the validation in transformer functions!

Reply:
:yay_fox: